### PR TITLE
Feature/wt 9718 tag property crud support open api spec

### PR DESCRIPTION
**Pull request by chanddan**

### CODE REVIEW

File: spec/openapi.yaml [hunk 1]
```diff
@@ -458,6 +514,40 @@ components:
...
+          type: string
+          description: The tag from which to delete the tag property set.
+        propertyNames:
+          description: The tag property to delete.
+          type: array
+          items:
+            type: string
+      required:
+        - tagName
+        - tagProperties
```
Comment (tom-haines) [inline spec/openapi.yaml]: Should be `propertyNames` ?

File: spec/openapi.yaml [hunk 2]
```diff
@@ -482,6 +572,18 @@ components:
...
             Use / to create tag in root.
           type: string
+        tagManagedBy:
+          description: >
+            Identifier of the resource that manages this tag.
+            If blank or missing, the value will be defaulted to the connector's name.
+          type: string
+        tagProperties:
+          description: >
+            Add one or more tag properties to the tag. The default behaviour ensures that there will be **no** duplicate or deletion of any existing tag specific properties within WiseTime.
```
Comment (tom-haines) [inline spec/openapi.yaml]:
```suggestion
            You can assign metadata to a tag.  You provide this optional information as a name-value (key-value) pair.  In the case of providing any name-value where the name is already assigned as metadata to the tag, the value will be updated with value supplied.
```

File: spec/openapi.yaml [hunk 3]
```diff
@@ -482,6 +572,18 @@ components:
...
+            Use the endpoint `/tag/properties/delete` for the deletion of tag properties.
```
Comment (tom-haines) [inline spec/openapi.yaml]:
```suggestion
            To delete/remove a name-value pair stored as metadata on a tag, you must use the `delete-tag-properties` operation.
```

File: spec/openapi.yaml [hunk 4]
```diff
@@ -482,6 +572,18 @@ components:
             For example /myteam/inprotech/
             Use / to create tag in root.
           type: string
+        tagManagedBy:
+          description: >
+            Identifier of the resource that manages this tag.
+            If blank or missing, the value will be defaulted to the connector's name.
+          type: string
+        tagProperties:
```
Comment (tom-haines) [inline spec/openapi.yaml]: Suggest we follow naming convention of metadata, which looks more common name for this type of key-value map (others call it a 'tag' like AWS S3 API, which would be quite confusing in our context).

a. As we reuse same object in `AddSetTagPropertiesRequest` also, could we schema ref to a `TagMetadataSet` ?  We could then use schema ref also in schema rep of Tag itself as needed.

b. Assuming type is TagMetadataSet, could we then `tagProperties` to simply `metadata` ? 

c. Think might be clearer if we use `/tag/metadata` in place of `/tag/properties` as a tag has other properties, such as last modified timestamp and createdBy etc.

Property

File: spec/openapi.yaml [hunk 5]
```diff
@@ -458,6 +514,40 @@ components:
...
+    AddSetTagPropertiesResponse:
+      type: object
+    DeleteTagPropertiesRequest:
+      type: object
+      properties:
+        tagName:
+          type: string
+          description: The tag from which to delete the tag property set.
+        propertyNames:
+          description: The tag property to delete.
```
Comment (tom-haines) [inline spec/openapi.yaml]:
```suggestion
          description: The list of tag property names to delete.
```

File: spec/openapi.yaml [hunk 6]
```diff
@@ -458,6 +514,40 @@ components:
...
+        tagName:
+          type: string
+          description: The name of the tag that will be contain the tag property.
+        tagProperties:
+          type: object
+          description: A set of tag properties.
+          additionalProperties:
+            type: string
+          example:
+            PRIMARY_GROUP: billable/non-billable
```
Comment (tom-haines) [inline spec/openapi.yaml]: we should use lowercase in example, and mention in MetadataSet (see other comment) that all tag property names are converted/stored in lowercase.

**Review by tom-haines — CHANGES_REQUESTED** [review]
Have made some initial suggestions re naming and schema ref reuse (not sure if we can or not)

**Comment by chanddan** [comment]
@tom-haines I have updated the spec, based on your naming suggestion I have renamed "tag properties" to "metadata". Agreed, it's a more descriptive name for its intent.

Also added TagMetadataSet.

File: spec/openapi.yaml [hunk 7]
```diff
@@ -120,6 +120,65 @@ paths:
         '500':
           description: An unexpected error occured.
 
+  /tag/metadata:
+    post:
+      operationId: add-set-tag-metadata
```
Comment (tom-haines) [inline spec/openapi.yaml]: for clarity, suggest `set-tag-metadata` for this operation id

File: spec/openapi.yaml [hunk 8]
```diff
@@ -120,6 +120,65 @@ paths:
...
+        You can assign metadata to a tag.  This information is provided as a name-value (key-value) pair.  In the case of providing any name-value where the name is already assigned as metadata to the tag, the value will be updated with value supplied.
+        To delete/remove a name-value pair stored as metadata on a tag, you must use the `delete-tag-metadata` operation.
+      tags:
+        - Tags
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/UpdateTagMetadataRequest'
```
Comment (tom-haines) [inline spec/openapi.yaml]: small nit - if operation called `Assign` then schema object should use same verb (replacing of `Update`), or vice-versa

File: spec/openapi.yaml [hunk 9]
```diff
@@ -120,6 +120,65 @@ paths:
         '500':
           description: An unexpected error occured.
 
+  /tag/metadata:
+    post:
+      operationId: assign-tag-metadata
```
Comment (tom-haines) [inline spec/openapi.yaml]: nit: other parts of api use highest concept to lowest concept semantics, to this should be

`tag-metadata-udpate`, `TagMetadataUpdateRequest` etc

**Review by tom-haines — CHANGES_REQUESTED** [review]
as discussed, let's follow concept ordering semantics for consistency

**Comment by chanddan** [comment]
@tom-haines  I have resolved your nits.

**Review by tom-haines — APPROVED** [review]

**Review by vyshane — APPROVED** [review]

File: spec/openapi.yaml [hunk 10]
```diff
@@ -120,6 +120,39 @@ paths:
...
+      responses:
+        '200':
+          description: Tag metadata successfully deleted.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TagMetadataDeleteResponse'
+        '204':
+          description: >
+            (No Content) when there is no need to return any content, where metadata was not found or if it was already deleted.
```
Comment (tom-haines) [inline spec/openapi.yaml]: Sorry to say but this design is still problematic, as we allow multiple `metadataNames` to be provided in the delete request, but only allow for a unary response.  I suggest we discuss to find clean solution.

File: spec/openapi.yaml [hunk 11]
```diff
@@ -120,6 +120,39 @@ paths:
...
+      responses:
+        '200':
+          description: Tag metadata successfully deleted.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TagMetadataDeleteResponse'
+        '204':
+          description: >
+            (No Content) metadata was not found or already deleted.
```
Comment (vyshane) [inline spec/openapi.yaml]:
```suggestion
```

@chanddan I didn't realise that we were deleting multiple tag metadata entries in one request. If that is the case then it seems simpler and less confusing to always return 200 OK when tag was found and there was no error processing the request, whether metadata was deleted or not. My 204 suggestion doesn't seem all that great now. What do you think?

File: spec/openapi.yaml [hunk 12]
```diff
@@ -120,6 +120,39 @@ paths:
...
+        - Tags
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/TagMetadataDeleteRequest'
+      responses:
+        '200':
+          description: Tag metadata successfully deleted.
```
Comment (vyshane) [inline spec/openapi.yaml]:
```suggestion
          description: Tag metadata successfully deleted. This operation is idempotent, and will also return HTTP code 200 if some or all of the tag metadata names were not found.
```

If you agree with [this suggestion](https://github.com/wisetime-io/connect-api-spec/pull/29/files#r396850452), then perhaps something like the above makes sense.

File: spec/openapi.yaml (hunk 11 above)
Comment (chanddan) [reply spec/openapi.yaml]: I have added a reply in `connect-api-server/pull-requests/159/overview`

File: spec/openapi.yaml (hunk 12 above)
Comment (chanddan) [reply spec/openapi.yaml]: Removed 204 from spec.

**Review by tom-haines — APPROVED** [review]
On deployment of additional endpoint, this can be merged to master.

**Pull request merged**